### PR TITLE
feat(akita-ng-effects): add library akita-ng-effects

### PR DESCRIPTION
**Pull request by yackinn**
initial setup## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/datorama/akita/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

https://github.com/datorama/akita/issues/571

Issue Number: #571

## What is the new behavior?

Effects are included. Listen and act upon action events with rxjs operators.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
The compiler is showing two errors depending on the Angular version and compiler settings.

This store.ts probably needs an injectable decorator in ivy Angular versions.

Error: libs/akita/src/lib/store.ts:48:14 - error NG2007: Class is using Angular features but is not decorated. Please add an explicit Angular decorator.
    
    48 export class Store<S = any> {

In some circumstances I saw the error that no function calls should be used inside a decorator.
AkitaNgEffects.forRoot([]) shouldn't be any different than RouterModule.forRoot() if you compare them as function calls.

### CODE REVIEW

File: apps/angular-store-app/src/app/app.module.ts
Comment (NetanelBasal) [inline apps/angular-store-app/src/app/app.module.ts:33]: What do u mean by "An Angular update will fix this"?

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (NetanelBasal) [inline apps/angular-store-app/src/app/cart/state/cart.effects.ts:11]: remove console.log, please

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (NetanelBasal) [inline apps/angular-store-app/src/app/cart/state/cart.effects.ts:25]: Use the `createEffect` function in one of these methods, please.

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (NetanelBasal) [inline apps/angular-store-app/src/app/cart/state/cart.effects.ts:21]: I think it'll be better to follow the `ngrx/effects` style that does the following:
- By default, it expects the `effect` to return a new action and dispatches it internally
- You can pass `{ dispatch: false }` as the second argument to cancel the previous behavior.

File: libs/akita-ng-effects/README.md
Comment (NetanelBasal) [inline libs/akita-ng-effects/README.md:1]: Add it also to our official docs, please. You can find it in the `docs` folder.

File: libs/akita-ng-effects/README.md
Comment (NetanelBasal) [inline libs/akita-ng-effects/README.md:64]:
```suggestion
    switchMap(_ => this.navigationService.loadMainNavigation().pipe(
```

File: libs/akita-ng-effects/README.md
Comment (NetanelBasal) [inline libs/akita-ng-effects/README.md:65]:
```suggestion
      tap(mainNav => this.actions$.dispatch(loadMainNavigationSuccess({ mainNav }))))
```

File: libs/akita-ng-effects/README.md
Comment (NetanelBasal) [inline libs/akita-ng-effects/README.md:71]:
```suggestion
    ofType(loadMainNavigationSuccess),
```

File: libs/akita-ng-effects/README.md
Comment (NetanelBasal) [inline libs/akita-ng-effects/README.md:91]:
```suggestion
        if (!isInit) this.actions.dispatch(loadMainNavigation());
```

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/__tests__/actions.spec.ts:6]: Why we need `fakeAsync` here? It's sync operation.

File: libs/akita-ng-effects/src/__tests__/create-effect.util.spec.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/__tests__/create-effect.util.spec.ts:15]: Why we need `done` here? It's sync operation.

File: libs/akita-ng-effects/src/__tests__/effect.decorator.spec.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/__tests__/effect.decorator.spec.ts:9]: Why we need `done` here? It's sync operation.

File: libs/akita-ng-effects/src/lib/akita-ng-effects.module.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/lib/akita-ng-effects.module.ts:50]: Why we need `mergedEffects`? We can just call `createEffectInstances(injector, effects)`

**Review by NetanelBasal — CHANGES_REQUESTED** [review]
Great work!

See my comments.

One more thing. Do you think we can add support for registering effects at the component providers level? It'll be useful to free some memory when the component is destroyed, and we don't need to listen anymore.

File: apps/angular-store-app/src/app/app.module.ts
Comment (yackinn) [reply apps/angular-store-app/src/app/app.module.ts:33]: I tried to run the project with multiple Angular versions for the akita store app.
Angular 8 had thrown an error.
Ivy versions worked fine without issues.
I haven't seen this error before.
Why wouldn't it be possible to use a function call within a decorator?

I'm not sure how to deal with this.
We can update the projects Angular version. That would also take a few changes to the code base.
I tried that already.

Comment (NetanelBasal) [reply apps/angular-store-app/src/app/app.module.ts:33]: Ok, I will check it when you finish

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:6]: As far as I'm concerned this is async.
The subject publishes the event and the assertion should run inside the async callback of the subscription.
It results in a false positive once the fakeAsync is removed.
The assertion won't run.

File: libs/akita-ng-effects/src/__tests__/create-effect.util.spec.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/create-effect.util.spec.ts:15]: As far as I'm concerned this is async.
The test will finish before the assertion has run.
It results in a false positive once the fakeAsync is removed.

File: libs/akita-ng-effects/src/__tests__/effect.decorator.spec.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/effect.decorator.spec.ts:9]: As far as I'm concerned this is async.
The test will finish before the assertion inside the tap operator has run.
It results in a false positive once the fakeAsync is removed.
In this test the assertion inside the tap operator isn't crucial I guess.

File: libs/akita-ng-effects/src/lib/akita-ng-effects.module.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/lib/akita-ng-effects.module.ts:50]: Absolutely correct. That was a leftover from an earlier different approach.

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (yackinn) [reply apps/angular-store-app/src/app/cart/state/cart.effects.ts]: removed

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (yackinn) [reply apps/angular-store-app/src/app/cart/state/cart.effects.ts]: Good call.
When this pull request is merged I might have a few more feature adaptions for the effects.
We can chat later about this.

File: apps/angular-store-app/src/app/cart/state/cart.effects.ts
Comment (yackinn) [reply apps/angular-store-app/src/app/cart/state/cart.effects.ts]: Changed.

File: libs/akita-ng-effects/README.md
Comment (yackinn) [reply libs/akita-ng-effects/README.md:1]: It's included now.

**Comment by yackinn** [comment]
Yes we can add a feature to provide the effects on a component level. Let me check.

Do you have any suggestions for the implementation?

If the effect isn't provided in any module we need to make sure to unsubscribe from the effects when the component is destroyed.
Can we use a decorator for such an effect?

@EffectsCollection()
class TestEffects {
}

File: libs/akita-ng-effects/README.md
Comment (yackinn) [reply libs/akita-ng-effects/README.md:64]: changed

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: https://stackblitz.com/edit/angular-testing-qknhej?file=app/hello.component.spec.ts

File: libs/akita-ng-effects/src/lib/module-manager.service.ts
Comment (NetanelBasal) [inline libs/akita-ng-effects/src/lib/module-manager.service.ts:15]: Please add types to any argument in this file

**Comment by NetanelBasal** [comment]
@yackinn, you don't need a decorator. The `ngOnDestroy` hook will be called in this case.

**Comment by NetanelBasal** [comment]
I had another thought. There is really nothing specific to Akita in this library. I think it'll be better to move it to `ngneat/effects` to be used with Akita, plain services, or any other approach.

WDYT?

**Comment by yackinn** [comment]
> I had another thought. There is really nothing specific to Akita in this library. I think it'll be better to move it to `ngneat/effects` so that it can be used with Akita, plain services, or any other approach.
> 
> WDYT?

If you ask me effects should be part of state management. 
People that want to checkout Akita should get a sense for a complete solution wihout the need for installing some part from another project. 
It should also be documented in the Akita docs as effects belong to state management.
I also want to integrate it more with the redux dev tools.
For my taste it's too detached from the redux logs. Every action should be visible in the dev tools with their respective payload. It helps immensly for debugging.

What do you think?

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: I did the same thing in webstorm and it failed.
I'll fix it

**Comment by yackinn** [comment]
> @yackinn, you don't need a decorator. The `ngOnDestroy` hook will be called in this case.

I need to pass an instruction to the ngOnDestroy to unsubscrube from all subscriptions inside the class.
How would you implement unsubscribing internally when ngOnDestroy is triggered?

**Comment by NetanelBasal** [comment]
> What do you think?

Yes, I agree. Let us add the integration for the devtools and schematics.

**Comment by NetanelBasal** [comment]
I'll let you know about the component's providers issue. I'm busy right now, so let's leave it to the end.

**Comment by NetanelBasal** [comment]
@yackinn when you finish, please merge `v6` branch into yours.

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: The tests won't pass correctly using the synchronous approach.

Bildschirmfoto 2020-12-20 um 18 40 41

File: libs/akita-ng-effects/src/lib/module-manager.service.ts
Comment (yackinn) [reply libs/akita-ng-effects/src/lib/module-manager.service.ts:15]: Types added.

**Comment by yackinn** [comment]
> @yackinn when you finish, please merge `v6` branch into yours.

Branch is merged and pushed.

**Comment by yackinn** [comment]
I'm thinking about what's the best solution to implement the logging of actiions with their payload in the redux dev tools.

The current state is the action decorator won't take any payload that can be logged with the action type. We can take care of this.

I don't want to overcomplicate the api.

I might have a proposal.
The action decorator can take an Action object created by the createAction function. 
The logAction function can take a StoreSnapshotAction instead of a string.

We can get the payload from the function arguments of the method that the action decorator is attached to.

We shouldn't log anything before the "reducer" in the state has performed the state update.

That means we need one action decorator for each action that should be posted in the redux dev tools.

For default actions it can just be the type update and the payload from the method arguments.


What should happen if you dispatch an action in the effects?
Would you need to trigger the store update excplicitly again with calling a action annotated method inside the store?

I guess this needs some thought.


I already tried to implement the updated action decorator and getting the payload from the class method. Logging the action with properties or payload in the redux dev tools works smoothly.

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (NetanelBasal) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: It doesn't make sense to me. I'll check it after the merge.

**Comment by NetanelBasal** [comment]
I think we should provide a way to hook into the `akitaDevtools` function and dispatch custom actions. We can pass an optional `actions` stream that we can subscribe to and call `devTools.send`.

File: libs/akita-ng-effects/src/__tests__/actions.spec.ts
Comment (NetanelBasal) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: Which font are you using?

Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: Should the next function be invoked before subscribing and asserting?

Comment (yackinn) [reply libs/akita-ng-effects/src/__tests__/actions.spec.ts:5]: I created that monospaced font myself.

**Comment by yackinn** [comment]
That seems feasible.
What about the action decorator? 
There are many actions in the redux dev tools that only provide a generic type such as "update" without any payload.

That's not really usefull I guess.

**Comment by NetanelBasal** [comment]
You can trigger actions that don't update the state. In this case, you want to see the action and the payload.

**Comment by yackinn** [comment]
That's true. But that's just one use case.
The loadBooksSuccess should trigger a state update. With the current setup that woul trigger two actions in the dev tools.
1. type: loadBooksSuccess; payload: .../ state difference: -
2. type: update: payload: -	/ state difference: ...

**Comment by NetanelBasal** [comment]
You are right.

**Comment by NetanelBasal** [comment]
I think we should rethink the functionality of the actions in general. I don't mind to have breaking changes as we're going to release v6.

**Comment by yackinn** [comment]
I agree.
What about something like this?

// books.actions.ts
export const loadBooks = createAction("[Books] Load Books")
export const loadBooksSucceess = createAction("[Books] Load Books Success", props<{books: Books}>())

// books.effects.ts
@Effect()
loadBooks$ = this.actions.pipe(
	ofType(loadBooks),
	switchmap(_ => ...api.loadBooks().pipe(
		map(books => loadBooksSuccess({books}))
	))
)

// books.store.ts
@Reducer( loadBooksSuccess, ... )
updateBooks(books: Books) {
	this.store.update(state => {...state, books})
}

**Comment by NetanelBasal** [comment]
I don't want to mix Redux with Akita. We already have actions, and I prefer not to bring the concept of reducers. 

I think we should keep it simple and transparent for the consumer. It should be handle somehow in the core Actions class.

**Comment by NetanelBasal** [comment]
We can use https://github.com/datorama/akita/blob/master/libs/akita/src/lib/actions.ts#L22

**Comment by yackinn** [comment]
I already did that. It works.
```ts
// public API for custom actions. Custom action always wins
export function logAction(action: StoreSnapshotAction) {
  setAction(action);
  customActionActive = true;
}
```
We need to distinguish between different action types though.
The effect shouldn't be the spot where dev tools logging is triggered from.
I think triggering the dev tool update needs to be issued by the store when the state is update by a certain action.

**Comment by NetanelBasal** [comment]
https://discord.gg/GCKBpYnw

**Comment by yackinn** [comment]
Logging for effects is implemented. 
Please check.

**Pull request closed without merging**
